docs/developer/style_guide.md

```diff
@@ -0,0 +1,579 @@
+
+call foo(int_array, N1, N2)
+...
+
+subroutine foo(int_array, N1, N2)
+implicit none
+integer, intent(in) :: N1, N2
+integer, dimension(1:N1, 1:N2) :: int_array
+...
+
+```
+
+as this does not cause any copying of the array and allows the compiler to do
+bound-checking etc.
+
+The use of **assumed-shape arrays**, such as `int_array(:, :)` is less efficient
+as this has an explicit rank, but an implicit shape. This requires the use of
+functions such as `size(int_array,1)` and may prevent certain compiler
+optimisations. However, it does allow the compiler to do bound-checking.
+
+Care must be taken in passing the array though. If an array section is to be
+passed then it is much better to use **assumed-size arrays** (i.e. the old
+fashioned way), that is
+
+```fortran
+
+call foo2(int_array(1, N))
+...
+
+subroutine foo2(int_array)
+implicit none
+integer, dimension(*) :: int_array
+...
+
+```
+
+than to pass `int_array(:, N)` and receive `int_array(:)` as this will create a
+copy of the array on the stack which will rapidly lead to "out of memory"
+errors.
+
+The use of assumed-size arrays, such as `int_array(*)` is useful if you want to
+write a general purpose routine that will work with any size and/or rank
+array. Note however, that it will not in general work with scalars, nor does it
+allow bounds-checking. However, it does not cause the creation of any copies of
+the array.
+
+
+## Machine-Specific Code
+
+### Avoid!
+
+Write in standard f90 without vendor extensions and without using any system
+routines wherever possible, for maximum portability. The one exception that is
+allowed to this is in the `Utility` modules, where all the low-level
+input/output and communications are handled.
+
+
+### Test thoroughly to prevent
+
+Always test your routines on several different platforms and with several
+different compilers to be sure that you have not inadvertently relied upon some
+default behaviour of one specific compiler. The CI system will handle
+compilation for all supported compilers once submitted as a merge request into
+the Bitbucket repository, but these should be performed beforehand to minimise
+the effort in bugfixing.
+
+
+### File handling
+
+One obvious area of machine dependency is that of file handling. For this reason,
+the `io` module should be used to handle the opening of all files. The programmer
+is responsible for reading / writing the data and closing the file. Files are
+not to be kept open for any longer than strictly necessary.
+
+
+### Compiler/preprocessor directives
+
+These should be avoided wherever possible as they hinder portability. They should
+only be used in the `Utility` modules for machine specific code. The one exception
+to this rule, is the use of a single `#ifdef debug` block at the start
+of each routine to check for conditions of usage. For this reason, all module
+filenames should be given the extension `.F90` not `.f90` to enable
+preproccesing on the widest range of compilers available.
```

Comment:
No longer true?

Reply:
Agreed. This should say .F90 ONLY in Utility modules, and no preprocessor in Fundamental or Functional. We made many of the '#ifdef debug' tests unconditional on the grounds that there was little cost and lot of gain in having them wherever they were needed. This reduced the number of -D needed, and now only machine/OS ones in general (and CODATA versions etc).